pyfr/plugins/kernels/turbulence.mako

```diff
@@ -0,0 +1,113 @@
+# -*- coding: utf-8 -*-
+<%namespace module='pyfr.backends.base.makoutil' name='pyfr'/>
+
+<%pyfr:macro name='turbulence' params='t, u, ploc, src'>
+  fpdtype_t xin = 0.0;
+  fpdtype_t ls = ${ls};
+  fpdtype_t ls2 = ${ls*ls};
+  fpdtype_t invls2 = ${1.0/(ls*ls)};
+  fpdtype_t gc3 = ${gc*gc*gc};
+  fpdtype_t rootrs = ${rootrs};
+  fpdtype_t srafac = ${srafac};
+  fpdtype_t invsigma2 = ${1.0/(sigma*sigma)};
+  fpdtype_t invsigma3 = ${1.0/(sigma*sigma*sigma)};
+  fpdtype_t ubar = ${ubar};
+  fpdtype_t pos[${ndims}];
+  fpdtype_t ttploc[${ndims}];
+  fpdtype_t eps[${ndims}];
+  fpdtype_t delta2[${ndims}];
+  fpdtype_t arg;
+  fpdtype_t utilde[${ndims}];
+  utilde[0] = 0.0;
+  utilde[1] = 0.0;
+  utilde[2] = 0.0;
+  fpdtype_t xloc2;
+  fpdtype_t sclip;
+  fpdtype_t g;
+  fpdtype_t xmin = -${ls};
+  fpdtype_t xmax = ${ls};
+  fpdtype_t ymin = ${ymin};
+  fpdtype_t ymax = ${ymax};
+  fpdtype_t zmin = ${zmin};
+  fpdtype_t zmax = ${zmax};
+  fpdtype_t fac = ${-0.5/(sigma*sigma*ls*ls)};
+  fpdtype_t fac2 = ${(gc*gc*gc)/(sigma*sigma*sigma)};
+
+  fpdtype_t tbc = 2.3283064365386962890625e-10;
+  
+  unsigned int oldstate;
+  unsigned int newstate;
+  unsigned int rshift;
+  unsigned int b22 = 22;
+  unsigned int b32 = 32;
+  unsigned int opbits = 4;
+
+  int epscomp;
+  
+  % for i, r in enumerate(rot):
+    ttploc[${i}] = ${' + '.join(f'{r[j]}*(ploc[{j}] - {shift[j]})' for j in range(3))};
+  % endfor
+
+  % for i in range(nvmax):
+    pos[0] = xmin + (t-tinit[${i}][0])*ubar;
+
+    oldstate = state[${i}][0];
+    newstate = (oldstate * 747796405UL) + 2891336453UL;
+    rshift = oldstate >> (b32 - opbits);
+    oldstate ^= oldstate >> (opbits + rshift);
+    oldstate *= 277803737UL;
+    oldstate ^= oldstate >> b22;
+    pos[1] = ymin + (ymax-ymin)*((fpdtype_t)oldstate * tbc);
+    //pos[1] = ymin + (ymax-ymin)*ldexp((fpdtype_t)oldstate, -32);
+    
+    oldstate = newstate;
+    newstate = (oldstate * 747796405UL) + 2891336453UL;
+    rshift = oldstate >> (b32 - opbits);
+    oldstate ^= oldstate >> (opbits + rshift);
+    oldstate *= 277803737UL;
+    oldstate ^= oldstate >> b22;
+    pos[2] = zmin + (zmax-zmin)*((fpdtype_t)oldstate * tbc);
+    
+    oldstate = newstate;
+    newstate = (oldstate * 747796405UL) + 2891336453UL;
+    rshift = oldstate >> (b32 - opbits);
+    oldstate ^= oldstate >> (opbits + rshift);
+    oldstate *= 277803737UL;
+    oldstate ^= oldstate >> b22;
+    epscomp = oldstate % 8;
```

Comment:
We don't need this.  If we are going to shift by 8 bits to get the last floating point value then we've got 8 bits which are being thrown away.  Lets use them for epscomp (ideally bits 7, 6, and 5 since they should be better than 0, 1, and 2).

Reply:
So you mean do something like epscomp = oldstate and then just get the three bits for eps via (epscomp >> 29) & 1 and (epscomp >> 29) & 2 and (epscomp >> 29) & 4?

Reply:
Yes, on L69 we only make use of 24 of the bits in oldstate.  This means there are 8 bits left that we can use for the three epscomp values.  By extension, we can also get rid of L64 since we do not need another new state and everything from L71-L76.

Regards, Freddie.

Reply:
Thanks.

Reply:
Or we run the third pass and get a third oldstate and then use the top 3 bits of it for eps (i.e. if we bother to make the third oldstate then use the most random bits from it). That is probably the best way from randomness perspective?

Reply:
The state should be good for 32-bits.  We are only making use of 24-bits for the floating point number, so there is no reason to not make use of the remaining 8-bits (of which we need 3 for eps).